core/src/kura.rs

```diff
@@ -4,258 +4,475 @@
+                        blocks.push(decoded_block);
+                    }
+                    Err(error) => {
+                        error!(%error, "Encountered malformed block. Not reading any blocks beyond this height.");
+                        break;
+                    }
+                },
+                Err(error) => {
+                    error!(%error, "Malformed block index or corrupted block data file. Not reading any blocks beyond this height.");
+                    break;
+                }
+            }
+        }
+
+        info!("Loaded {} blocks at init.", blocks.len());
+
+        let hash_array = blocks.iter().map(VersionedCommittedBlock::hash).collect();
+
+        let mut guard = self
+            .block_hash_array
+            .lock()
+            .expect("lock on block hash array");
+        *guard = hash_array;
+
+        Ok(blocks)
     }
-}
 
-#[async_trait::async_trait]
-impl<IO: DiskIO> Actor for KuraWithIO<IO> {
-    fn actor_channel_capacity(&self) -> u32 {
-        self.actor_channel_capacity
+    // I think we should rethink this. Kura should not be the
+    // one to start other services. Kura should be a provider of
+    // block storage services only.
+    /// Call kura init then world state view init and start sumeragi.
+    /// # Panics
+    /// * if kura init fails.
+    #[allow(clippy::unwrap_used)]
+    pub async fn async_init_all_important(&self) {
+        let blocks = self.init().unwrap();
+        self.wsv.init(blocks).await;
+        let last_block = self.wsv.latest_block_hash();
+        let height = self.wsv.height();
+        self.broker
+            .issue_send(sumeragi::message::Init { last_block, height })
+            .await;
     }
 
-    async fn on_start(&mut self, ctx: &mut Context<Self>) {
-        self.broker.subscribe::<StoreBlock, _>(ctx);
+    #[allow(clippy::expect_used, clippy::cognitive_complexity, clippy::panic)]
+    fn kura_recieve_blocks_thread(weak_kura_reference: &Weak<Kura>) {
+        loop {
+            let strong_kura_reference = match weak_kura_reference.upgrade() {
+                Some(kura_arc) => kura_arc,
+                None => {
+                    info!("Kura block thread is shutting down");
+                    return;
+                }
+            };
+
+            let mut block_reciever_guard = strong_kura_reference
+                .block_reciever
+                .lock()
+                .expect("able to lock kura recieve block mutex");
 
-        #[allow(clippy::panic)]
-        match self.init().await {
-            Ok(blocks) => {
-                #[cfg(feature = "telemetry")]
-                if let Some(block) = blocks.first() {
-                    iroha_logger::telemetry!(msg = iroha_telemetry::msg::SYSTEM_CONNECTED, genesis_hash = %block.hash());
+            match block_reciever_guard.try_recv() {
+                Err(_) => {
+                    std::thread::sleep(std::time::Duration::from_millis(1));
```

Comment:
Keep in mind that thread sleep might release the OS thread, but not the run-time thread, which might need yield.

Reply:
Yes, rules are very different when in the runtime vs out of the runtime. Or if you are on the "blocking" tokio thread pool vs the normal one.